File: app/src/main/java/info/nukoneko/cuc/kidspos/setting/SettingFragment.java
package info.nukoneko.cuc.kidspos.setting;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.EditTextPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;

import info.nukoneko.cuc.kidspos.R;

/**
 * Created by atsumi on 2016/02/20.
 */
public class SettingFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.preference);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Preference p = getPreferenceScreen().findPreference(key);
//        if(p instanceof EditTextPreference)
//        {
//            setEdited(!((EditTextPreference) p).getText().equals(p.getSummary()));
//            setPrefData(key, ((EditTextPreference) p).getText());
//        }else if(p instanceof ListPreference){
//            setEdited(!((ListPreference) p).getValue().equals(p.getSummary()));
//            setPrefData(key, ((ListPreference) p).getValue());
//        }
    }
    @Override
    public void onResume() {
        super.onResume();
        // Set up a listener whenever a key changes
        getPreferenceScreen().getSharedPreferences()
                .registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        // Unregister the listener whenever a key changes
        getPreferenceScreen().getSharedPreferences()
                .unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = super.onCreateView(inflater, container, savedInstanceState);
        if(v != null) {
            ListView lv = (ListView) v.findViewById(android.R.id.list);
            lv.setPadding(0, 0, 0, 0);
        }
        return v;
    }
}

File: app/src/main/java/info/nukoneko/cuc/android/kidspos/ui/setting/SettingFragment.kt
package info.nukoneko.cuc.android.kidspos.ui.setting

import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.preference.EditTextPreference
import android.support.v7.preference.PreferenceFragmentCompat
import android.support.v7.preference.SwitchPreferenceCompat
import android.text.TextUtils

import info.nukoneko.cuc.android.kidspos.KidsPOSApplication
import info.nukoneko.cuc.android.kidspos.R
import info.nukoneko.cuc.android.kidspos.api.manager.SettingsManager
import info.nukoneko.cuc.android.kidspos.event.ApplicationEvent
import info.nukoneko.cuc.android.kidspos.util.AlertUtil
import info.nukoneko.cuc.android.kidspos.util.ValidationUtil

class SettingFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    override fun onCreatePreferences(savedInstanceState: Bundle, rootKey: String) {
        addPreferencesFromResource(R.xml.settings)
        updateSummaries()
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        val app = KidsPOSApplication[context] ?: return

        when (key) {
            SettingsManager.KEY_ENABLE_PRACTICE_MODE -> app.postEvent(ApplicationEvent.AppModeChange)
            SettingsManager.KEY_SERVER_IP -> {
                val ip = sharedPreferences.getString(key, "")
                if (TextUtils.isEmpty(ip) || !ValidationUtil.isValidIPv4Address(ip!!)) {
                    AlertUtil.showErrorDialog(context!!, "IPアドレスが間違っています", false, null)
                    sharedPreferences.edit().putString(key, SettingsManager.DEFAULT_IP).apply()
                    return
                }
            }
            SettingsManager.KEY_SERVER_PORT -> {
                val port = sharedPreferences.getInt(key, -1)
                if (!ValidationUtil.isUsablePort(port)) {
                    AlertUtil.showErrorDialog(context!!, "ポートが間違っています", false, null)
                    sharedPreferences.edit().putInt(key, SettingsManager.DEFAULT_PORT).apply()
                    return
                }
            }
        }
        updateSummaries()
    }

    private fun updateSummaries() {
        run {
            val pref = findPreference(SettingsManager.KEY_SERVER_IP) as EditTextPreference
            pref.summary = pref.text
        }
        run {
            val pref = findPreference(SettingsManager.KEY_SERVER_PORT) as EditTextPreference
            pref.summary = pref.text
        }
        run {
            val pref = findPreference(SettingsManager.KEY_ENABLE_PRACTICE_MODE) as SwitchPreferenceCompat
            pref.summary = if (pref.isChecked) "練習モード" else "通常モード"
        }
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        super.onPause()
    }

    companion object {
        fun newInstance(): SettingFragment = SettingFragment()
    }
}
